File: lambda/src/main/java/model/test/TestInstance.java
package model.test;

import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBHashKey;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBRangeKey;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTable;

import java.io.Serializable;
import java.util.List;

@DynamoDBTable(tableName = "TestInstances")
public class TestInstance implements Serializable {
    private String applicantId;
    private long timestamp;
    private String title;
    private List<SolvableClosedQuestion> closeQuestions;
    private List<SolvableOpenQuestion> openQuestions;
    private List<SolvableValueQuestion> valueQuestions;
    private float maxScore;
    private float receivedScore;
    private int status;
    private String recruiterId;
    private Long testId;

    public TestInstance() {
    }

    public TestInstance(String applicantId, long timestamp, String title, List<SolvableClosedQuestion> closeQuestions, List<SolvableOpenQuestion> openQuestions, List<SolvableValueQuestion> valueQuestions, float maxScore, float receivedScore, int status) {
        this.applicantId = applicantId;
        this.timestamp = timestamp;
        this.title = title;
        this.closeQuestions = closeQuestions;
        this.openQuestions = openQuestions;
        this.valueQuestions = valueQuestions;
        this.maxScore = maxScore;
        this.receivedScore = receivedScore;
        this.status = status;
    }

    @DynamoDBHashKey(attributeName = "applicantId")
    public String getApplicantId() {
        return applicantId;
    }

    public void setApplicantId(String applicantId) {
        this.applicantId = applicantId;
    }

    @DynamoDBRangeKey(attributeName = "timestamp")
    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }


    public List<SolvableClosedQuestion> getCloseQuestions() {
        return closeQuestions;
    }

    public void setCloseQuestions(List<SolvableClosedQuestion> closeQuestions) {
        this.closeQuestions = closeQuestions;
    }


    public List<SolvableOpenQuestion> getOpenQuestions() {
        return openQuestions;
    }

    public void setOpenQuestions(List<SolvableOpenQuestion> openQuestions) {
        this.openQuestions = openQuestions;
    }


    public List<SolvableValueQuestion> getValueQuestions() {
        return valueQuestions;
    }

    public void setValueQuestions(List<SolvableValueQuestion> valueQuestions) {
        this.valueQuestions = valueQuestions;
    }


    public float getMaxScore() {
        return maxScore;
    }

    public void setMaxScore(float maxScore) {
        this.maxScore = maxScore;
    }


    public float getReceivedScore() {
        return receivedScore;
    }

    public void setReceivedScore(float receivedScore) {
        this.receivedScore = receivedScore;
    }


    public int getStatus() {
        return status;
    }

    public void setStatus(int status) {
        this.status = status;
    }

    public Long getTestId() {
        return testId;
    }

    public void setTestId(Long testId) {
        this.testId = testId;
    }

    public String getRecruiterId() {
        return recruiterId;
    }

    public void setRecruiterId(String recruiterId) {
        this.recruiterId = recruiterId;
    }

    public void calculatePoints() {
        for (SolvableClosedQuestion c : this.closeQuestions) {
            this.receivedScore += c.getReceivedScore();
        }

        for (SolvableOpenQuestion c : this.openQuestions) {
            this.receivedScore += c.getReceivedScore();
        }

        for (SolvableValueQuestion c: this.valueQuestions){
            this.receivedScore += c.getReceivedScore();
        }
    }
}

File: web/src/app/model/test-instance.ts
import { SolvableOpenQuestion } from './solvable-open-question';
import { SolvableCloseQuestion } from './solvable-close-question';
import { SolvableValueQuestion } from './solvable-value-question';

export class TestInstance {
  applicantId: string;
  timestamp: number;
  title: string;
  status: TestStatus;
  openQuestions: SolvableOpenQuestion[];
  closeQuestions: SolvableCloseQuestion[];
  valueQuestions: SolvableValueQuestion[];
  maxScore: number;
  receivedScore: number;

  constructor();
  constructor(
    applicantId: string,
    timestamp: number,
    title: string,
    status: TestStatus,
    openQuestions: SolvableOpenQuestion[],
    closeQuestions: SolvableCloseQuestion[],
    valueQuestions: SolvableValueQuestion[],
    maxScore: number,
    receivedScore: number
    );
  constructor(
    applicantId?: string,
    timestamp?: number,
    title?: string,
    status?: TestStatus,
    openQuestions?: SolvableOpenQuestion[],
    closeQuestions?: SolvableCloseQuestion[],
    valueQuestions?: SolvableValueQuestion[],
    maxScore?: number,
    receivedScore?: number
    ) {
    this.applicantId = applicantId;
    this.timestamp = timestamp;
    this.title = title;
    this.status = status;
    this.openQuestions = openQuestions;
    this.closeQuestions = closeQuestions;
    this.valueQuestions = valueQuestions;
    this.maxScore = maxScore;
    this.receivedScore = receivedScore;
  }
}

export enum TestStatus {
  NotSolved,
  Solved,
  Checked
}
